docs/architecture.md

```diff
@@ -0,0 +1,331 @@
+}
+
+trait TransactionPool {
+    // remove a set of txs from the pool after a block is committed
+    fn drop_txs(txs: Vec<TxId>) -> Result<()>;
+}
+
+trait Database {
+    // insert consensus information associated with a sealed block
+    fn insert_consensus_data(sealed_block: &SealedFuelBlock) -> Result<()>;
+}
+
+```
+#### Adapters: fuel_core::service::adapters::block_importer
+```rust
+impl fuel_poa_consensus::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::BlockImporter for Service<BlockImporter> {
+    // insert impl here
+}
+```
+
+### Synchronizer
+
+#### Ports: fuel_sync::ports
+```rust
+trait PeerToPeer {
+    async fn fetch_best_peer_height() -> Result<(PeerId, BlockHeight)>;
+    async fn fetch_block_headers(query: Range<BlockHeight>) -> Result<Vec<(PeerId, SealedFuelBlockHeader)>>;
+    async fn fetch_block_data(id: block_id) -> Result<Vec<(PeerId, FuelBlockBody)>>;
+    async fn report_invalid_height(peer: PeerId) -> Result<()>;
+    async fn report_invalid_block(peer: PeerId) -> Result<()>;
+}
+
+trait BlockImporter {
+    // commit a sealed block to the chain
+    async fn commit(block: SealedFuelBlock) -> Result<()>;
+}
+
+trait Consensus {
+    // check with the consensus layer whether a block header passes consensus rules
+    async fn validate_sealed_block_header(block: SealedFuelBlockHeader) -> Result<()>;
+}
+
+trait Database {
+    // get current fuel blockchain height
+    fn get_current_height() -> Result<BlockHeight>;
+}
+```
+#### Adapters: fuel_core::service::adapters::fuel_sync
+_N/A_
+
+### P2P
+
+#### Ports: fuel_p2p::ports
+```rust
+trait Database {
+    // used when other peers are syncing
+    fn get_sealed_block(height: BlockHeight) -> Result<SealedFuelBlock>; 
+}
+```
+
+#### Adapters: fuel_core::service::adapters::peer_to_peer
+```rust
+impl fuel_poa_consensus::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_sync::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+impl fuel_txpool::ports::PeerToPeer for Service<P2pService> {
+    // insert impl here
+}
+
+```
+
+### Transaction Pool
+
+#### Ports: fuel_txpool::ports
+```rust
+trait PeerToPeer {
+    async fn broadcast_transaction(transaction: Transaction) -> Result<()>;
+    fn report_invalid_tx(peer_id: PeerId) -> Result<()>;
+    // handler for gossipped transactions
+    fn on_new_tx(tx: Transaction, peer: PeerId);
+}
+```
+
+#### Adapters: fuel_core::service::adapters::transaction_pool
+```rust
+impl block_importer::ports::TransactionPool for Service<TransactionPool> {
+    // insert impl here
+}
+
+impl block_producer::ports::TransactionPool for Service<TransactionPool> {
+    // insert impl here
+}
+```
+
+### Executor
+
+#### Ports: fuel_block_executor::ports
+```rust
+trait Database: IntepreterStorage {
+    fn get_coin(id: UtxoId) -> Result<Coin>;
+    fn upsert_coin(coin: Coin) -> Result<()>;
```

Comment:
What exactly does `upsert` mean? Is this to refer to replacing an old UTXO which has now been spent with a new one?

Reply:
Upsert means update/insert, meaning that the value is inserted if it doesn't exist, and updated if it does.